rbd: include trashed parent images while calculating the clone depth

nixpanic opened the pull request:
The `getCloneDepth()` function did not account for images that are in
the trash. A trashed image can only be opened by the image-id, and not
by name anymore.

Closes: #4013
Depends-on: #4064 #4273

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.27

Rakshith-R commented on internal/rbd/rbd_util.go:
```diff
@@ -697,47 +697,69 @@ func (ri *rbdImage) trashRemoveImage(ctx context.Context) error {
+		}
 
-			return depth, err
+		// of there is a parent, count it to the depth
```
```suggestion
		// if there is a parent, count it to the depth
```

Rakshith-R reviewed (commented):
This pr looks good to me.

nixpanic committed:
rbd: include trashed parent images while calculating the clone depth

The `getCloneDepth()` function did not account for images that are in
the trash. A trashed image can only be opened by the image-id, and not
by name anymore.

Closes: #4013
Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.27

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.27

nixpanic commented:
Manual stress testing with the scripts from rook/rook#12312 passes.

Need to fix the golangci-lint issues and have the go-ceph rebase merged before this is completely ready.

Rakshith-R commented on internal/rbd/rbd_util.go:
```diff
@@ -697,47 +697,58 @@ func (ri *rbdImage) trashRemoveImage(ctx context.Context) error {
+//
+// This function re-uses the ioctx of the image to open all images in the
+// chain. There is no need to open new ioctx's for every image.
+func (ri *rbdImage) getCloneDepth() (uint, error) {
```
hey @nixpanic,
Can we add another arguement like `maxDepthToTraverse` which can be used to exit and not traverse further unnecessarily? 

We can pass hard `rbdHardMaxCloneDepth+1` to the function call then.

wdyt?

nixpanic replied:
sure, I'll add that

nixpanic replied:
Added a 2nd commit to address this.

nixpanic committed:
rbd: pass a max-depth to `getCloneDepth()`

The `getCloneDepth()` function does not need to traverse the whole chain
of parents when a certain max-limit is configured. The traversing can be
aborted once the hard-limit is reached. This makes the procedure a
little more efficient, as unnecessary traversing is prevented.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

riya-singhal31 commented on internal/rbd/rbd_util.go:
```diff
@@ -697,47 +697,67 @@ func (ri *rbdImage) trashRemoveImage(ctx context.Context) error {
+// images in the chain. The `max` argument to the function is used to specify a
+// limit of the depth to check. When `max` depth is reached, no further
+// traversing of the depth is required.
+//
```
Maybe we can remove this "//" from here.

nixpanic replied:
I've done that intentionally. An empty line between the short description of the function, and a details about the implementation.

riya-singhal31 reviewed (dismissed):
LGTM, thanks Niels

nixpanic commented:
@mergifyio rebase

Madhu-1 reviewed (commented):
LGTM,  i will let @Rakshith-R to approve it as had some suggestions.

Rakshith-R reviewed (dismissed):
LGTM, thanks !

Rakshith-R commented:
@mergifyio refresh

Rakshith-R commented:
@mergifyio queue

Madhu-1 commented:
/test ci/centos/upgrade-tests-rbd

Madhu-1 commented:
/test ci/centos/mini-e2e-helm/k8s-1.28

Madhu-1 commented:
/test ci/centos/mini-e2e/k8s-1.26

Madhu-1 commented:
/test ci/centos/mini-e2e/k8s-1.27

nixpanic commented:
It seems that there was some issue with snapshots. not sure why e2e passed earlier?!

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.28

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -697,47 +697,65 @@ func (ri *rbdImage) trashRemoveImage(ctx context.Context) error {
+
+		// open the parent image, so that the for-loop can continue
+		// with checking for the parent of the parent
+		image, err = librbd.OpenImageById(ri.ioctx, info.Image.ImageID, info.Snap.SnapName)
```
Now passing `info.Snap.SnapName` instead of `librbd.NoSnapshot`.

nixpanic commented:
e2e without the last `go fmt` fix is available at https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.28/detail/mini-e2e-helm_k8s-1.28/78/pipeline . This PR can be reviewed once that passed.

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.28

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.28

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.28

nixpanic committed:
rbd: use `librbd.OpenImageById()` if `rbdVol.ImageID` is set

`librbd.OpenImageById()` works if the image is in the trash, so it makes
it possible to get the parent of the image.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.28

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.28

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.28

nixpanic commented:
For some reason, these commits prevent setting the meta-data for a cloned image (e2e suggests). When testing manually, no RBD-images get any meta-data set at all. The OMAP for all images (including the created snapshot) does have the expected meta-data... Still trying to understand what the e2e tests expect, and why they do not fail on previous test-cases.

nixpanic commented:
> For some reason, these commits prevent setting the meta-data for a cloned image (e2e suggests). When testing manually, no RBD-images get any meta-data set at all. The OMAP for all images (including the created snapshot) does have the expected meta-data... Still trying to understand what the e2e tests expect, and why they do not fail on previous test-cases.

It seemed that my deployment did not include `--setmetadata=true` for the provisioner... When setting it, the volume gets the expected (I think) metadata set. Trying again in the CI, now with enhanced logging.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.28

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.28

nixpanic commented:
For anyone watching this, I have now noticed that the ImageID of a `rbdSnapshot` is not set to the right value, it is set to the ImageID of the original parent volume. Using the ImageID to open the `rbdSnapshot`, actually opens the parent image.

nixpanic commented:
Dropped the commit

    rbd: use librbd.OpenImageById() if rbdVol.ImageID is set

as it seems that the ImageID is not always set to the correct value, but to the ImageID of a cloned image. Without this commit, e2e should pass.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.28

Madhu-1 commented on internal/csi-common/server.go:
```diff
@@ -97,7 +97,7 @@ func (s *nonBlockingGRPCServer) serve(endpoint string, srv Servers) {
 	}
 
-	server := grpc.NewServer()
+	server := grpc.NewServer(NewMiddlewareServerOption(false))
```
@nixpanic can we make a new PR from this commit and get it merged? not sure somehow we missed how it got removed

nixpanic replied:
sure!

nixpanic replied:
See #4263. Will drop it from this PR once the e2e run finishes.

nixpanic committed:
rbd: set/get correct ImageID in more places

In some places the ImageID is used as the ID of the parent. That is very
confusing and prone to errors. Instead, fetch the right ImageID where
possible, and set ParentID for referencing to parent images.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.28

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.28

nixpanic committed:
rbd: the DeleteVolume CSI procedure should succeed on deleted images

If the RBD-image is deleted already, the DeleteVolume CSI procedure is
expected to report success (as it should be idempotent). In case the
returned error indicates "RBD image not found", the error is ignored and
the DeleteVolume procedure continues.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
[logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.28/detail/mini-e2e_k8s-1.28/288/pipeline)

```
  I1127 15:20:04.710772       1 utils.go:164] ID: 39 Req-ID: 0001-0024-65e89782-f2bd-4b30-85f4-fcad649cdded-0000000000000001-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 GRPC call: /csi.v1.Controller/DeleteVolume
  I1127 15:20:04.710860       1 utils.go:165] ID: 39 Req-ID: 0001-0024-65e89782-f2bd-4b30-85f4-fcad649cdded-0000000000000001-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 GRPC request: {"secrets":"***stripped***","volume_id":"0001-0024-65e89782-f2bd-4b30-85f4-fcad649cdded-0000000000000001-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6"}
  I1127 15:20:04.712116       1 omap.go:88] ID: 39 Req-ID: 0001-0024-65e89782-f2bd-4b30-85f4-fcad649cdded-0000000000000001-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 got omap values: (pool="replicapool", namespace="", name="csi.volume.ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6"): map[csi.imageid:1af286243815 csi.imagename:csi-vol-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 csi.volname:pvc-3d33943d-80f0-46d1-ad6e-d6165ee1b7e5 csi.volume.owner:rbd-6238]
  I1127 15:20:04.757621       1 rbd_util.go:637] ID: 39 Req-ID: 0001-0024-65e89782-f2bd-4b30-85f4-fcad649cdded-0000000000000001-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 rbd: delete csi-vol-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6-temp using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789, pool replicapool
  I1127 15:20:04.762284       1 controllerserver.go:1021] ID: 39 Req-ID: 0001-0024-65e89782-f2bd-4b30-85f4-fcad649cdded-0000000000000001-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 deleting image csi-vol-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6
  I1127 15:20:04.762309       1 rbd_util.go:637] ID: 39 Req-ID: 0001-0024-65e89782-f2bd-4b30-85f4-fcad649cdded-0000000000000001-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 rbd: delete csi-vol-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789, pool replicapool
  E1127 15:20:04.763705       1 rbd_util.go:667] ID: 39 Req-ID: 0001-0024-65e89782-f2bd-4b30-85f4-fcad649cdded-0000000000000001-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 failed to delete rbd image: replicapool/csi-vol-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6, error: RBD image not found
  E1127 15:20:04.763736       1 controllerserver.go:1023] ID: 39 Req-ID: 0001-0024-65e89782-f2bd-4b30-85f4-fcad649cdded-0000000000000001-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 failed to delete rbd image: replicapool/csi-vol-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 with error: RBD image not found
  E1127 15:20:04.763801       1 utils.go:169] ID: 39 Req-ID: 0001-0024-65e89782-f2bd-4b30-85f4-fcad649cdded-0000000000000001-ccfe21c4-54cf-44cc-9a75-72ce1ad9adf6 GRPC error: rpc error: code = Internal desc = RBD image not found
```

Deleting the RBD image seems to have resulted in a (spurious?) error. The `DeleteVolume` CSI procedure is repeated over and over again, because the image was removed already.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.28

nixpanic committed:
rbd: skip flattening if an image in trash

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: return CSI expected errors while flattening a snapshot or volume

By returned `ABORTED` or `PRECONDITION_FAILED` in the right places, the
CO will retry with the same arguments until the snapshot is
`ReadyToUse`. This causes restoring a volume from a snapshot to be
delayed, until the snapshot can be used.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.27

nixpanic commented:
From the [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e_k8s-1.27/618/display/redirect) it looks like the flattening of images is in progress. I have not found messages that suggest something is blocked, but it is well possible that there is a deadlock on the PVC-restore and snapshot-delete somewhere (I have seen these in manual testing too, but though I addressed them).

Maybe https://github.com/ceph/ceph-csi/issues/1883#issuecomment-881253782 contains more hints on things that I have missed in the current PR.

l0nax commented:
@nixpanic do the problems still exist if rebased against `devel`?

I'm facing some issues with my Ceph Cluster. The images won't get flattened because the parent has been moved to the trash. It seems like this PR would address the problem. – I'm happy to contribute/ help if needed.

nixpanic committed:
rbd: log a warning when closing an image fails

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.31

nixpanic commented:
@Rakshith-R, do you think this might have been fixed? I tried to reproduce the issue with an RBD-image chain of 20 volumes (loop with: create, write data, snapshot, clone, delete, ...). Even before applying this PR for testing, there isn't an issue anymore :thinking:
